File: linkgame-server/src/main/java/io/github/octcarp/linkgame/server/net/ClientHandlerThread.java
package io.github.octcarp.linkgame.server.net;

import io.github.octcarp.linkgame.common.module.GridPos;
import io.github.octcarp.linkgame.common.module.Player;
import io.github.octcarp.linkgame.common.packet.*;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.net.Socket;
import java.net.SocketException;
import java.util.List;


public class ClientHandlerThread implements Runnable {

    private final Socket socket;
    private Player player;
    private final ObjectInputStream ois;
    private final ObjectOutputStream oos;

    public ClientHandlerThread(Socket accept) {
        this.socket = accept;
        try {
            oos = new ObjectOutputStream(socket.getOutputStream());
            ois = new ObjectInputStream(socket.getInputStream());
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void run() {
        while (true) {
            try {
                Object oin = ois.readObject();
                if (!(oin instanceof Request request)) {
                    throw new RuntimeException("Invalid request");
                }
                RequestType typeIn = request.getType();
                switch (typeIn) {
                    case LOGIN -> {
                        Player player = (Player) request.getData();
                        SimpStatus status = PlayersManager.getInstance().playerCanLogin(player);

                        if (status == SimpStatus.OK) {
                            this.player = player;
                            PlayersManager.getInstance().addPlayerThread(player.id(), this);
                        }

                        Response response = new Response(ResponseType.LOGIN_RESULT);
                        response.setData(status);
                        sendResponse(response);
                    }
                    case REGISTER -> {
                        Player newPlayer = (Player) request.getData();
                        SimpStatus status = PlayersManager.getInstance().registerPlayer(newPlayer);

                        Response response = new Response(ResponseType.REGISTER_RESULT);
                        response.setData(status);
                        sendResponse(response);
                    }
                    case LOGOUT -> {
                        SimpStatus status = PlayersManager.getInstance().playerLogout(player.id());

                        Response response = new Response(ResponseType.LOGOUT_RESULT);
                        response.setData(status);
                        sendResponse(response);
                    }
                    case ENTER_LOBBY -> {
                        LobbyManager.getInstance().enterLobby(player.id(), this);
                    }
                    case START_WAITING -> {
                        LobbyManager.getInstance().startWaiting(player.id());
                    }
                    case STOP_WAITING -> {
                        LobbyManager.getInstance().stopWaiting(player.id());
                    }
                    case EXIT_LOBBY -> {
                        LobbyManager.getInstance().exitLobby(player.id());
                    }
                    case JOIN_PLAYER -> {
                        String oppId = (String) request.getData();
                        LobbyManager.getInstance().joinPlayer(player.id(), oppId);
                    }
                    default -> {
                        if (player == null) {
                            Response response = new Response(ResponseType.ERROR_MESSAGE);
                            response.setData("Please login first");
                            sendResponse(response);
                        } else {
                            switch (typeIn) {
                                case SELECT_BOARD -> {
                                    GridPos pos = (GridPos) request.getData();
                                    MatchManager.getInstance().selectBoard(player.id(), pos);
                                }
                                case SHUFFLE_BOARD -> {
                                    MatchManager.getInstance().shuffleBoard(player.id());
                                }
                                case TURN_MOVE -> {
                                    List<GridPos> list = (List<GridPos>) request.getData();
                                    MatchManager.getInstance().judgeTurnMove(player.id(), list.getFirst(), list.getLast());
                                }
                                default -> {
                                    // send response to client
                                }
                            }
                        }
                    }
                }
            } catch (IOException e) {
                if (e instanceof SocketException) {
                    if (player != null) {
                        PlayersManager.getInstance().playerLogout(player.id());
                        LobbyManager.getInstance().playerDisconnected(player.id());
                        PlayersManager.getInstance().playerDisconnect(player.id());
                        MatchManager.getInstance().playerDisconnected(player.id());
                    }
                    System.out.println("Client:" + socket.getRemoteSocketAddress() + " disconnected");
                    try {
                        socket.close();
                    } catch (IOException ex) {
                        throw new RuntimeException(ex);
                    }
                }
                throw new RuntimeException(e);
            } catch (ClassNotFoundException e) {
                throw new RuntimeException(e);
            }
        }
    }

    public void sendResponse(Response response) {
        try {
            oos.writeObject(response);
            oos.flush();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: game-server/src/main/kotlin/io/github/octcarp/sustech/cs209a/linkgame/server/net/ClientHandlerThread.kt
package io.github.octcarp.sustech.cs209a.linkgame.server.net

import io.github.octcarp.sustech.cs209a.linkgame.common.model.GridPos
import io.github.octcarp.sustech.cs209a.linkgame.common.model.Player
import io.github.octcarp.sustech.cs209a.linkgame.common.packet.*
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream
import java.net.Socket

class ClientHandlerThread(private val socket: Socket) : Runnable {
    private var player: Player? = null
    private var ois: ObjectInputStream? = null
    private var oos: ObjectOutputStream? = null

    init {
        try {
            oos = ObjectOutputStream(socket.getOutputStream())
            ois = ObjectInputStream(socket.getInputStream())
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    override fun run() {
        while (true) {
            try {
                val oin = ois!!.readObject()
                if (oin !is Request) {
                    throw RuntimeException("Invalid request")
                }
                val typeIn = oin.type
                when (typeIn) {
                    RequestType.LOGIN -> {
                        val player = oin.data as Player
                        val status = PlayersManager.playerCanLogin(player)

                        if (status == SimpStatus.OK) {
                            this.player = player
                            PlayersManager.addPlayerThread(player.id, this)
                        }

                        val response = Response(ResponseType.LOGIN_RESULT)
                        response.data = status
                        sendResponse(response)
                    }

                    RequestType.REGISTER -> {
                        val newPlayer = oin.data as Player
                        val status = PlayersManager.registerPlayer(newPlayer)

                        val response = Response(ResponseType.REGISTER_RESULT)
                        response.data = status
                        sendResponse(response)
                    }


                    RequestType.GET_MATCH_RECORD -> {
                        val matches = RecordManager.matchRecordList
                        val response = Response(ResponseType.GET_MATCH_RECORD_RESULT)
                        response.data = matches
                        sendResponse(response)
                    }

                    RequestType.SHUTDOWN -> {
                        handleClientDisconnect()
                    }

                    else -> {
                        player?.let { player ->
                            when (typeIn) {
                                RequestType.LOGOUT -> {
                                    val status = PlayersManager.playerLogout(player.id)

                                    val response = Response(ResponseType.LOGOUT_RESULT)
                                    response.data = status
                                    sendResponse(response)
                                }

                                RequestType.ENTER_LOBBY -> {
                                    LobbyManager.enterLobby(player.id, this)
                                }

                                RequestType.START_WAITING -> {
                                    LobbyManager.startWaiting(player.id)
                                }

                                RequestType.STOP_WAITING -> {
                                    LobbyManager.stopWaiting(player.id)
                                }

                                RequestType.SELECT_BOARD -> {
                                    val pos = oin.data as GridPos
                                    MatchManager.selectBoard(player.id, pos)
                                }

                                RequestType.SHUFFLE_BOARD -> {
                                    MatchManager.shuffleBoard(player.id)
                                }

                                RequestType.EXIT_LOBBY -> {
                                    LobbyManager.exitLobby(player.id)
                                }

                                RequestType.JOIN_PLAYER -> {
                                    val oppId = oin.data as String
                                    LobbyManager.joinPlayer(player.id, oppId)
                                }

                                RequestType.RECONNECT_MATCH -> {
                                    val hasMatch = MatchManager.reconnectMatch(player.id)
                                    if (!hasMatch) {
                                        val response = Response(ResponseType.NO_MATCH_TO_RECONNECT)
                                        sendResponse(response)
                                    }
                                }

                                RequestType.EXIT_MATCH -> {
                                    MatchManager.playerDisconnected(player.id)
                                }

                                RequestType.TURN_MOVE -> {
                                    val (start, end) = (oin.data as List<GridPos>).let { it.first() to it.last() }
                                    MatchManager.judgeTurnMove(player.id, start, end)
                                }

                                else -> {}
                            }
                        } ?: run {
                            val response = Response(ResponseType.ERROR_MESSAGE)
                            response.data = "Please login first"
                            sendResponse(response)
                        }
                    }
                }
            } catch (_: IOException) {
                handleClientDisconnect()
                break
            } catch (e: ClassNotFoundException) {
                throw RuntimeException(e)
            }
        }

        cleanUp()
    }

    fun sendResponse(response: Response) {
        try {
            oos!!.writeObject(response)
            oos!!.flush()
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    private fun handleClientDisconnect() {
        println("Client:" + socket.getRemoteSocketAddress() + " disconnected")

        player?.let { player ->
            with(PlayersManager) {
                playerLogout(player.id)
                playerDisconnect(player.id)
            }
            LobbyManager.playerDisconnected(player.id)
            MatchManager.playerDisconnected(player.id)
        }
    }

    private fun cleanUp() {
        try {
            ois?.close()
            oos?.close()
            socket.close()
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }
}
